go/storage/mkvs/db/pathbadger/metadata.go

```diff
@@ -0,0 +1,193 @@
+package pathbadger
+
+import (
+	"fmt"
+	"math"
+	"sync"
+
+	"github.com/dgraph-io/badger/v4"
+
+	"github.com/oasisprotocol/oasis-core/go/common"
+	"github.com/oasisprotocol/oasis-core/go/common/cbor"
+	"github.com/oasisprotocol/oasis-core/go/storage/mkvs/db/api"
+)
+
+// maxPendingVersions is the maximum number of allowed non-finalized versions. Increasing this too
+// much can result in the metadata growing too much.
+const maxPendingVersions = 5000
+
+// serializedMetadata is the on-disk serialized metadata.
+type serializedMetadata struct {
+	// Version is the database schema version.
+	Version uint64 `json:"version"`
+	// Namespace is the namespace this database is for.
+	Namespace common.Namespace `json:"namespace"`
+
+	// EarliestVersion is the earliest version.
+	EarliestVersion uint64 `json:"earliest_version"`
+	// LastFinalizedVersion is the last finalized version.
+	LastFinalizedVersion *uint64 `json:"last_finalized_version"`
+	// MultipartVersion is the version for the in-progress multipart restore, or 0 if none was in
+	// progress.
+	MultipartVersion uint64 `json:"multipart_version"`
+	// MultipartSeqs are the sequence numbers used for the multipart restore.
+	MultipartSeqs map[uint8]uint16 `json:"multipart_seqs,omitempty"`
+
+	// NextPendingRootSeq contains the next pending root sequence number for a given type in the
+	// given version.
+	NextPendingRootSeq map[uint64]map[uint8]uint16 `json:"next_pending_root_seq,omitempty"`
+	// PendingRootSeqs contains the set of all non-finalized roots in the next version.
+	PendingRootSeqs map[uint64]map[api.TypedHash]uint16 `json:"pending_root_seqs,omitempty"`
+}
+
+// metadata is the database metadata.
+type metadata struct {
+	sync.RWMutex
+
+	value serializedMetadata
+}
+
+func (m *metadata) getEarliestVersion() uint64 {
+	m.RLock()
+	defer m.RUnlock()
+
+	return m.value.EarliestVersion
+}
+
+func (m *metadata) setEarliestVersion(tx *badger.Txn, version uint64) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// The earliest version can only increase, not decrease.
+	if version < m.value.EarliestVersion {
+		return fmt.Errorf("earliest version must only increase")
+	}
+
+	m.value.EarliestVersion = version
+	return m.save(tx)
+}
+
+func (m *metadata) getLastFinalizedVersion() (uint64, bool) {
+	m.RLock()
+	defer m.RUnlock()
+
+	if m.value.LastFinalizedVersion == nil {
+		return 0, false
+	}
+	return *m.value.LastFinalizedVersion, true
+}
+
+func (m *metadata) setLastFinalizedVersion(tx *badger.Txn, version uint64) error {
+	m.Lock()
+	defer m.Unlock()
+
+	if m.value.LastFinalizedVersion != nil && version <= *m.value.LastFinalizedVersion {
```

Comment:
Hm, not sure why this is here as we already do the check you mention in `Finalize`. Probably here it would only make sense to panic if `version != *m.value.LastFinalizedVersion + 1`.

Reply:
Oh I see, this actually wouldn't work as we allow arbitrary changes in case of multipart restore over a previous database. The current implementation is ok, maybe we just need a comment.

Reply:
Or we could disallow multipart restore into an existing database :-)